addons/controllers/kapp-controller/kappcontrollerconfig_controller.go

```diff
@@ -0,0 +1,219 @@
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, fmt.Sprintf("unable to fetch cluster %s/%s", clusterNamespacedName.Namespace, clusterNamespacedName.Name))
+		return ctrl.Result{}, err
+	}
+
+	if retResult, err := r.ReconcileKappControllerConfig(ctx, kappControllerConfig, cluster, log); err != nil {
+		log.Error(err, "unable to reconcile kappControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KappControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&runv1alpha3.KappControllerConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+// ReconcileKappControllerConfig reconciles KappControllerConfig CR
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfig(
+	ctx context.Context,
+	kappControllerConfig *runv1alpha3.KappControllerConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger) (_ ctrl.Result, retErr error) {
+
+	var (
+		patchConfig bool
+	)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(kappControllerConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Patch KappControllerConfig before returning the function
+	defer func() {
+		// patchConfig will be true if ownerrefence or secretRef is added or deleted
+		if patchConfig {
+			log.Info("Patching kappControllerConfig")
+
+			if err := patchHelper.Patch(ctx, kappControllerConfig); err != nil {
+				log.Error(err, "Error patching kappControllerConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	if err := r.ReconcileKappControllerConfigNormal(ctx, kappControllerConfig, cluster, log, &patchConfig); err != nil {
+		log.Error(err, "Error reconciling kappControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Successfully reconciled kappControllerConfig")
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileKappControllerConfigNormal reconciles KappControllerConfig CR
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfigNormal(
+	ctx context.Context,
+	kappControllerConfig *runv1alpha3.KappControllerConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger,
+	patchConfig *bool) (retErr error) {
+
+	// add owner reference to kappControllerConfig
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               cluster.Kind,
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		BlockOwnerDeletion: pointer.BoolPtr(false),
+	}
+
+	if !clusterapiutil.HasOwnerRef(kappControllerConfig.OwnerReferences, ownerReference) {
+		log.Info("Adding owner reference to kappControllerConfig")
+		kappControllerConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(kappControllerConfig.OwnerReferences, ownerReference)
+		*patchConfig = true
```

Comment:
This is not needed but in general mutating a parameter to indicate that something was done or not is anti-pattern. If this is needed the function should return a boolean instead.

Reply:
Removed the check